### PR TITLE
Refactor trainers to use list of obs rather than vec and vis obs

### PR DESCRIPTION
**Pull request by vincentpierre**
### Proposed change(s)

Instead of separating visual and vector observations in the buffer and the trainer, we have a single list of observations.

 - This is the "flat buffer" option from this [document](https://docs.google.com/document/d/1Btq__Ooo0RfXCrTEQDCxcJwG_tiiu5nBtu0CRP3zSw0/edit)

 - This change is incompatible with `Barracuda <= 1.1.2-preview`

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [x] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/policy/torch_policy.py [hunk 1]
```diff
@@ -95,31 +95,29 @@ def export_memory_size(self) -> int:
...
-    def update_normalization(self, vector_obs: np.ndarray) -> None:
+    def update_normalization(self, buffer: AgentBuffer) -> None:
         """
         If this policy normalizes vector observations, this will update the norm values in the graph.
         :param vector_obs: The vector observations to add to the running estimate of the distribution.
         """
-        vector_obs = [torch.as_tensor(vector_obs)]
         if self.use_vec_obs and self.normalize:
-            self.actor_critic.update_normalization(vector_obs)
+            self.actor_critic.update_normalization(buffer)
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/policy/torch_policy.py:115]: Not sure how I feel about passing the whole buffer (rather than just the obs) to the normalizer. But probably doesn't really matter

Comment (vincentpierre) [reply ml-agents/mlagents/trainers/policy/torch_policy.py:116]: The way the observations as list is made, there can be multiple vector observations per Agent, this allows to normalize potentially all of the observations that can be normalized.

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 2]
```diff
@@ -290,8 +280,25 @@ def forward(
         At this moment, torch.onnx.export() doesn't accept None as tensor to be exported,
         so the size of return tuple varies with action spec.
         """
+        # This code will convert the ugly vec and obs into glorious unified list of inputs
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/torch/networks.py:283]: At some point we should change the C# side too. But today is not this day.

File: ml-agents/mlagents/trainers/policy/torch_policy.py [hunk 3]
```diff
@@ -98,59 +98,56 @@ def export_memory_size(self) -> int:
 
     def _split_decision_step(
         self, decision_requests: DecisionSteps
-    ) -> Tuple[SplitObservations, np.ndarray]:
-        vec_vis_obs = SplitObservations.from_observations(decision_requests.obs)
+    ) -> Tuple[List[np.array], np.ndarray]:
+        obs = decision_requests.obs
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/policy/torch_policy.py:102]: This feels weird now. Whatever calls _split_decision_step should just get the `obs` from `DecisionSteps` instead.

Comment (chriselion) [reply ml-agents/mlagents/trainers/policy/torch_policy.py:102]: and _split_decision_step should just return the mask

File: ml-agents/mlagents/trainers/policy/torch_policy.py [hunk 4]
```diff
@@ -98,59 +98,56 @@ def export_memory_size(self) -> int:
...
-            vec_obs, vis_obs, masks, memories, seq_len
+            obs, masks, memories, seq_len
         )
         return (actions, log_probs, entropies, memories)
 
     def evaluate_actions(
         self,
-        vec_obs: List[torch.Tensor],
-        vis_obs: List[torch.Tensor],
+        obs: torch.Tensor,
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/policy/torch_policy.py:143]: `List[torch.Tensor]`?

File: ml-agents/mlagents/trainers/policy/torch_policy.py [hunk 5]
```diff
@@ -164,19 +161,17 @@ def evaluate(
         :param decision_requests: DecisionStep object containing inputs.
         :return: Outputs from network as defined by self.inference_dict.
         """
-        vec_vis_obs, masks = self._split_decision_step(decision_requests)
-        vec_obs = [torch.as_tensor(vec_vis_obs.vector_observations)]
-        vis_obs = [
-            torch.as_tensor(vis_ob) for vis_ob in vec_vis_obs.visual_observations
-        ]
+        obs, masks = self._split_decision_step(decision_requests)
+        obs = [torch.as_tensor(np_ob) for np_ob in obs]
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/policy/torch_policy.py:165]: maybe `np_obs` and `tensor_obs` to make the types clearer (instead of redefining `obs`)?

File: ml-agents/mlagents/trainers/torch/components/bc/module.py [hunk 6]
```diff
@@ -137,12 +138,17 @@ def _behavioral_cloning_loss(
...
         """
         Helper function for update_batch.
         """
-        vec_obs = [ModelUtils.list_to_tensor(mini_batch_demo["vector_obs"])]
+        obs = ObsUtil.from_buffer(
+            mini_batch_demo, len(self.policy.behavior_spec.observation_shapes)
+        )
+        # Convert to tensors
+        obs = [ModelUtils.list_to_tensor(obs) for obs in obs]
+        print("\n\n\n\n", obs, obs[0].shape)
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/torch/components/bc/module.py:151]: remove print

File: ml-agents/mlagents/trainers/torch/components/reward_providers/curiosity_reward_provider.py [hunk 7]
```diff
@@ -106,38 +107,24 @@ def get_current_state(self, mini_batch: AgentBuffer) -> torch.Tensor:
...
-            ],
-            vis_inputs=[
-                ModelUtils.list_to_tensor(
-                    mini_batch["visual_obs%d" % i], dtype=torch.float
-                )
-                for i in range(n_vis)
-            ],
-        )
+        n_obs = len(self._state_encoder.encoders)
+        obs = ObsUtil.from_buffer(mini_batch, n_obs)
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/torch/components/reward_providers/curiosity_reward_provider.py:111]: Same comment about redefining `obs`. If mypy ever gets smart enough to understand the types it'll complain about this (currently it thinks they're both List[Any])

File: ml-agents/mlagents/trainers/trajectory.py [hunk 8]
```diff
@@ -18,48 +18,40 @@ class AgentExperience(NamedTuple):
...
-class SplitObservations(NamedTuple):
-    vector_observations: np.ndarray
-    visual_observations: List[np.ndarray]
+class ObsUtil:
+    @staticmethod
+    def get_name_at(index: int) -> str:
+        """
+        returns the name of the observation given the index of the observation
+        """
+        return "obs_%d" % index
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/trajectory.py:27]: How about about tuple instead of string concatenation? 
```
$ python3 -m timeit '{("obs", n): n for n in range(10000)}'
500 loops, best of 5: 718 usec per loop
$ python3 -m timeit '{f"obs_{n}": n for n in range(10000)}'
200 loops, best of 5: 1.77 msec per loop
$ python3 -m timeit '{f"obs_%d" % n: n for n in range(10000)}'
200 loops, best of 5: 1.84 msec per loop
```
If that doesn't convince you, at least use an f-string.

**Review by chriselion — APPROVED** [review]
Looks good overall

File: ml-agents/mlagents/trainers/trajectory.py (hunk 8 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/trajectory.py]: I use an f-string now

File: ml-agents/mlagents/trainers/policy/torch_policy.py (hunk 4 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/policy/torch_policy.py]: my mistake

File: ml-agents/mlagents/trainers/policy/torch_policy.py (hunk 3 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/policy/torch_policy.py]: I renamed the method and it now only returns the masks

File: ml-agents/mlagents/trainers/torch/utils.py [hunk 9]
```diff
@@ -115,57 +115,61 @@ def _check_resolution_for_encoder(
...
+        vis_encode_type: EncoderType,
+    ) -> Tuple[nn.Module, int]:
+        """
+        Returns the encoder and the size of the appropriate encoder.
+        :param shape: Tuples that represent the observation dimension.
+        :param normalize: Normalize all vector inputs.
+        :param h_size: Number of hidden units per layer.
+        :param vis_encode_type: Type of visual encoder to use.
+        """
+        if len(shape) == 1:
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/torch/utils.py:132]: Eventually, this will use the DimensionProperties to decide an encoder correct?

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 10]
```diff
@@ -58,39 +58,34 @@ def __init__(
...
-            processed_vec = processor(vec_input)
-            encodes.append(processed_vec)
-
-        for idx, processor in enumerate(self.visual_processors):
-            vis_input = vis_inputs[idx]
-            if not exporting_to_onnx.is_exporting():
-                vis_input = vis_input.permute([0, 3, 1, 2])
-            processed_vis = processor(vis_input)
-            encodes.append(processed_vis)
+        for idx, processor in enumerate(self.encoders):
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/torch/networks.py:85]: Should we call this variable `self.processors`? To differentiate between these and the `LinearEncoder` that's further down in the NetworkBody

**Review by ervteng — APPROVED** [review]
Minor comment about naming of `self.encoders` but otherwise LGTM

File: ml-agents/mlagents/trainers/torch/components/reward_providers/curiosity_reward_provider.py [hunk 11]
```diff
@@ -106,38 +107,24 @@ def get_current_state(self, mini_batch: AgentBuffer) -> torch.Tensor:
...
-                ModelUtils.list_to_tensor(
-                    mini_batch["next_visual_obs%d" % i], dtype=torch.float
-                )
-                for i in range(n_vis)
-            ],
-        )
+        n_obs = len(self._state_encoder.encoders)
+        obs = ObsUtil.from_buffer_next(mini_batch, n_obs)
+        # Convert to tensors
+        obs = [ModelUtils.list_to_tensor(obs) for obs in obs]
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/torch/components/reward_providers/curiosity_reward_provider.py:125]: Sorry, I didn't call out all of the places that redefine `obs` - here's another one, and there are several others.

File: ml-agents/mlagents/trainers/policy/torch_policy.py (hunk 1 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/policy/torch_policy.py:116]: I'm thinking going forward we can replace this with a data structure (e.g. ObservationsTuple) that would be created from the buffer, and can pass in here. But this is OK for this PR

**Review by chriselion — APPROVED** [review]

File: ml-agents/mlagents/trainers/torch/components/reward_providers/curiosity_reward_provider.py (hunk 11 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/components/reward_providers/curiosity_reward_provider.py:125]: I will do another pass

File: ml-agents/mlagents/trainers/torch/utils.py (hunk 9 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/utils.py:132]: Yes

File: ml-agents/mlagents/trainers/torch/networks.py (hunk 10 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/networks.py:85]: Will do
